### PR TITLE
[Fix-10827] Fix network error cause worker cannot send message to master

### LINKED ISSUE #10827: [Improvement][RPC] When client send request to server use a separate connection
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

Right now, the master and worker will send request/response to each other. And we use async mode, so once the client send a request to the server, he will not block, he just cache the request, and use a daemon thread to retry, if he doesn't receive the ack of the server in a while. Once he receive the server's ack, he will remove the related cached request.
image

The problem is that if the server send result to client by the old channel, the server may failed, due to the old channel closed.

e.g.

1. Master dispatch task to worker, the master nettyClient port is a random port(11111), the worker nettyServer port is 1234.
2. Worker will cache the channel(masterHost:11111).
3. Worker will send response to master by the cached channel(masterHost:11111) and will cache the response in memory used to retry if he doesn't received the ack of his response.

If the channel masterHost: 11111 is closed, then the worker will always send response to master failed, since master doesn't has a nettyServer listen the 11111 port.

So I think when worker need to sent something to master, he need to send to the master's nettyServer, not master's nettyClient, if the channel closed, he can still create a new channel. So the worker need to know the master's nettyServer port, we need to add this in worker's config.
image

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by ruanwenjun**
## Purpose of the pull request

close #10827 
- ~~Rename some command to message~~, add sourceAddress and targetAddress in message
- When worker send message to master use the separate connection
- Add BaseMessage class, our  message body should extends this class to set sourceAddress and target Address.

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

### CODE REVIEW

**Comment by ruanwenjun** [comment]
@caishunfeng This PR is ready to review, please take a look.

**Comment by caishunfeng** [comment]
Hi @ruanwenjun I found many changes in this pr is about `command to message`, is it necessary to change it? 
I think we can keep the command name to reduce modification and cherry-pick costs, WDYT?

**Comment by ruanwenjun** [comment]
> Hi @ruanwenjun I found many changes in this pr is about `command to message`, is it necessary to change it?
> I think we can keep the command name to reduce modification and cherry-pick costs, WDYT?

I add a new class `BaseMessage`, and our async command will extend this class, to get some fields.
I suggestion to rename command to message, since we  already have a Command for netty, and the other command class is only the command body.

**Comment by ruanwenjun** [comment]
I will revert the name change related to command.

**Comment by ruanwenjun** [comment]
@caishunfeng I have reverted the rename change of command, please take a look.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/processor/TaskExecuteRunningProcessor.java [hunk 1]
```diff
@@ -54,10 +54,10 @@ public class TaskExecuteRunningProcessor implements NettyRequestProcessor {
     @Override
     public void process(Channel channel, Command command) {
         Preconditions.checkArgument(CommandType.TASK_EXECUTE_RUNNING == command.getType(), String.format("invalid command type : %s", command.getType()));
-        TaskExecuteRunningCommand taskExecuteRunningCommand = JSONUtils.parseObject(command.getBody(), TaskExecuteRunningCommand.class);
-        logger.info("taskExecuteRunningCommand: {}", taskExecuteRunningCommand);
+        TaskExecuteRunningCommand taskExecuteRunningMessage = JSONUtils.parseObject(command.getBody(), TaskExecuteRunningCommand.class);
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/processor/TaskExecuteRunningProcessor.java:57]: nip: revert variable name

File: dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/processor/TaskAckProcessorTest.java [hunk 2]
```diff
@@ -44,7 +44,7 @@ public class TaskAckProcessorTest {
     private TaskExecuteRunningProcessor taskExecuteRunningProcessor;
     private TaskEventService taskEventService;
     private ProcessService processService;
-    private TaskExecuteRunningCommand taskExecuteRunningCommand;
+    private TaskExecuteRunningCommand taskExecuteRunningMessage;
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/processor/TaskAckProcessorTest.java:47]: same here

File: dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/message/MessageSender.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package org.apache.dolphinscheduler.server.worker.message;
+
+import org.apache.dolphinscheduler.plugin.task.api.TaskExecutionContext;
+import org.apache.dolphinscheduler.remote.command.BaseCommand;
+import org.apache.dolphinscheduler.remote.command.CommandType;
+import org.apache.dolphinscheduler.remote.exceptions.RemotingException;
+
+public interface MessageSender<T extends BaseCommand> {
```
Comment (caishunfeng) [inline dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/message/MessageSender.java:25]: :+1:

File: dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/processor/TaskCallbackService.java [hunk 4]
```diff
@@ -76,201 +59,201 @@ public TaskCallbackService() {
...
-
-    /**
-     * send task execute response command
-     */
-    public void sendRecallCommand(TaskExecutionContext taskExecutionContext) {
-        TaskRecallCommand taskRecallCommand = buildRecallCommand(taskExecutionContext);
-        ResponseCache.get().cache(taskExecutionContext.getTaskInstanceId(), taskRecallCommand.convert2Command(), TaskEventType.WORKER_REJECT);
-        send(taskExecutionContext.getTaskInstanceId(), taskRecallCommand.convert2Command());
-    }
+    //    /**
```
Comment (caishunfeng) [inline dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/processor/TaskCallbackService.java]: maybe it should be removed.

File: dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/rpc/WorkerMessageSender.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    @PostConstruct
+    public void init() {
+        Map<String, MessageSender> messageSenders = applicationContext.getBeansOfType(MessageSender.class);
+        messageSenders.values().forEach(messageSender -> messageSenderMap.put(messageSender.getMessageType(),
+                                                                              messageSender));
+    }
+
+    // todo: use message rather than context
+    public void sendMessageNeedAck(@NonNull TaskExecutionContext taskExecutionContext,
```
Comment (caishunfeng) [inline dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/rpc/WorkerMessageSender.java]:
```suggestion
    public void sendMessageWithRetry(@NonNull TaskExecutionContext taskExecutionContext,
```

**Review by caishunfeng — COMMENTED** [review]
LGTM overall, some nip comments.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/processor/TaskExecuteRunningProcessor.java (hunk 1 above)
Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/processor/TaskExecuteRunningProcessor.java:57]: Done

File: dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/processor/TaskCallbackService.java (hunk 4 above)
Comment (ruanwenjun) [reply dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/processor/TaskCallbackService.java]: Done

File: dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/rpc/WorkerMessageSender.java (hunk 5 above)
Comment (ruanwenjun) [reply dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/rpc/WorkerMessageSender.java]: Done

**Review by caishunfeng — APPROVED** [review]
+1

**Pull request merged**